torch/nn/modules/instancenorm.py

```diff
@@ -36,29 +36,90 @@ def eval(self):
         return self
 
 
+class LayerNorm(_InstanceNorm):
+    r"""Applies Layer Normalization over a 2D input that is seen
+    as a mini-batch of 1D inputs.
+
+    .. math::
+
+        y = \gamma * \frac{x - \mu_x}{\sigma_x + \epsilon} + \beta
+
+    The mean and standard deviation are calculated per-dimension separately
+    for each object in a mini-batch. Gamma and beta are learnable parameter
+    vectors of size C (where C is the input size).
+
+    During training, this layer keeps a running estimate of its computed mean
+    and variance. The running sum is kept with a default momentum of 0.1.
+
+    At evaluation time (`.eval()`), the default behaviour of the LayerNorm
+    module stays the same, i.e. the running mean/variance is NOT used for
+    normalization. One can force using the stored mean and variance with
+    the `.train(False)` method.
```

Comment:
Agreed, that code looked mysterious to me and I wasn't sure it worked either. Pinging @DmitryUlyanov to find out more. As James said below, using running statistics seems to be counter to layer/instance norm, and if so, then we could actually change InstanceNorm to use what I wrote for LayerNorm (I'll keep the instance norm name, just update the docs to refer to layer norm too): https://github.com/pytorch/pytorch/pull/2019